## Exploring the relationship between gender and policing

In [1]:
import pandas as pd

In [2]:
pd.options.display.max_rows=200

In [3]:
source_path = 'F:/Data Camp Courses/Analyzing Police Activity with Pandas/Exploring the relationship between gender and policing/'

In [4]:
source_file = 'ri_cleansed.csv'

In [5]:
ri = pd.read_csv(source_path + source_file,
                parse_dates=['stop_date','stop_datetime'],
                 index_col ='stop_datetime'
                )

In [6]:
ri.shape

(86536, 13)

In [7]:
ri.head()

stop_date stop_time driver_gender driver_race  \
stop_datetime                                                        
2005-01-04 12:55:00 2005-01-04     12:55             M       White   
2005-01-23 23:15:00 2005-01-23     23:15             M       White   
2005-02-17 04:15:00 2005-02-17     04:15             M       White   
2005-02-20 17:15:00 2005-02-20     17:15             M       White   
2005-02-24 01:20:00 2005-02-24     01:20             F       White   

                                      violation_raw  violation  \
stop_datetime                                                    
2005-01-04 12:55:00  Equipment/Inspection Violation  Equipment   
2005-01-23 23:15:00                        Speeding   Speeding   
2005-02-17 04:15:00                        Speeding   Speeding   
2005-02-20 17:15:00                Call for Service      Other   
2005-02-24 01:20:00                        Speeding   Speeding   

                     search_conducted search_type   stop_outcome  is_arrested  \
stop_datetime                                                                   
2005-01-04 12:55:00             False         NaN       Citation        False   
2005-01-23 23:15:00             False         NaN       Citation        False   
2005-02-17 04:15:00             False         NaN       Citation        False   
2005-02-20 17:15:00             False         NaN  Arrest Driver         True   
2005-02-24 01:20:00             False         NaN       Citation        False   

                    stop_duration  drugs_related_stop district  
stop_datetime                                                   
2005-01-04 12:55:00      0-15 Min               False  Zone X4  
2005-01-23 23:15:00      0-15 Min               False  Zone K3  
2005-02-17 04:15:00      0-15 Min               False  Zone X4  
2005-02-20 17:15:00     16-30 Min               False  Zone X1  
2005-02-24 01:20:00      0-15 Min               False  Zone X3

Examining traffic violations

Before comparing the violations being committed by each gender, you should examine the violations committed by all drivers to get a baseline understanding of the data.

In this exercise, you'll count the unique values in the violation column, and then separately express those counts as proportions.


    Count the unique values in the violation column, to see what violations are being committed by all drivers.
    Express the violation counts as proportions of the total.



In [8]:
ri['violation'].value_counts()

Speeding               48423
Moving violation       16224
Equipment              10921
Other                   4409
Registration/plates     3703
Seat belt               2856
Name: violation, dtype: int64

In [9]:
ri['violation'].value_counts(normalize=True)

Speeding               0.559571
Moving violation       0.187483
Equipment              0.126202
Other                  0.050950
Registration/plates    0.042791
Seat belt              0.033004
Name: violation, dtype: float64

Comparing violations by gender

The question we're trying to answer is whether male and female drivers tend to commit different types of traffic violations.

In this exercise, you'll first create a DataFrame for each gender, and then analyze the violations in each DataFrame separately.


    Create a DataFrame, female, that only contains rows in which driver_gender is 'F'.
    Create a DataFrame, male, that only contains rows in which driver_gender is 'M'.
    Count the violations committed by female drivers and express them as proportions.
    Count the violations committed by male drivers and express them as proportions.


In [10]:
female = ri[ri['driver_gender']=='F']

In [11]:
male  = ri[ri['driver_gender']=='M']

In [12]:
female['violation'].value_counts(normalize=True)

Speeding               0.658114
Moving violation       0.138218
Equipment              0.105199
Registration/plates    0.044418
Other                  0.029738
Seat belt              0.024312
Name: violation, dtype: float64

In [13]:
male['violation'].value_counts(normalize=True)

Speeding               0.522243
Moving violation       0.206144
Equipment              0.134158
Other                  0.058985
Registration/plates    0.042175
Seat belt              0.036296
Name: violation, dtype: float64

Comparing speeding outcomes by gender

When a driver is pulled over for speeding, many people believe that gender has an impact on whether the driver will receive a ticket or a warning. Can you find evidence of this in the dataset?

First, you'll create two DataFrames of drivers who were stopped for speeding: one containing females and the other containing males.

Then, for each gender, you'll use the stop_outcome column to calculate what percentage of stops resulted in a "Citation" (meaning a ticket) versus a "Warning".


    Create a DataFrame, female_and_speeding, that only includes female drivers who were stopped for speeding.
    Create a DataFrame, male_and_speeding, that only includes male drivers who were stopped for speeding.
    Count the stop outcomes for the female drivers and express them as proportions.
    Count the stop outcomes for the male drivers and express them as proportions.


In [14]:
female_and_speeding = ri[(ri.driver_gender=='F') & (ri.violation=='Speeding')]

In [15]:
female_and_speeding.shape

(15646, 13)

In [16]:
male_and_speeding = ri[(ri.driver_gender=='M') & (ri.violation=='Speeding')]

In [17]:
male_and_speeding.shape

(32777, 13)

In [18]:
female_and_speeding.head()

stop_date stop_time driver_gender driver_race  \
stop_datetime                                                        
2005-02-24 01:20:00 2005-02-24     01:20             F       White   
2005-03-14 10:00:00 2005-03-14     10:00             F       White   
2005-07-14 11:20:00 2005-07-14     11:20             F       White   
2005-07-18 19:30:00 2005-07-18     19:30             F       White   
2005-07-24 20:10:00 2005-07-24     20:10             F       White   

                    violation_raw violation  search_conducted search_type  \
stop_datetime                                                               
2005-02-24 01:20:00      Speeding  Speeding             False         NaN   
2005-03-14 10:00:00      Speeding  Speeding             False         NaN   
2005-07-14 11:20:00      Speeding  Speeding             False         NaN   
2005-07-18 19:30:00      Speeding  Speeding             False         NaN   
2005-07-24 20:10:00      Speeding  Speeding             False         NaN   

                    stop_outcome  is_arrested stop_duration  \
stop_datetime                                                 
2005-02-24 01:20:00     Citation        False      0-15 Min   
2005-03-14 10:00:00     Citation        False      0-15 Min   
2005-07-14 11:20:00     Citation        False      0-15 Min   
2005-07-18 19:30:00     Citation        False      0-15 Min   
2005-07-24 20:10:00     Citation        False      0-15 Min   

                     drugs_related_stop district  
stop_datetime                                     
2005-02-24 01:20:00               False  Zone X3  
2005-03-14 10:00:00               False  Zone K3  
2005-07-14 11:20:00               False  Zone X4  
2005-07-18 19:30:00               False  Zone K3  
2005-07-24 20:10:00               False  Zone K3

In [19]:
print(female_and_speeding['stop_outcome'].value_counts(normalize=True))

Citation            0.952192
Warning             0.040074
Arrest Driver       0.005752
N/D                 0.000959
Arrest Passenger    0.000639
No Action           0.000383
Name: stop_outcome, dtype: float64


In [20]:
print(male_and_speeding['stop_outcome'].value_counts(normalize=True))

Citation            0.944595
Warning             0.036184
Arrest Driver       0.015895
Arrest Passenger    0.001281
No Action           0.001068
N/D                 0.000976
Name: stop_outcome, dtype: float64


## Interesting! The numbers are similar for males and females: about 95% of stops for speeding result in a ticket. Thus, the data fails to show that gender has an impact on who gets a ticket for speeding.

Calculating the search rate

During a traffic stop, the police officer sometimes conducts a search of the vehicle. In this exercise, you'll calculate the percentage of all stops that result in a vehicle search, also known as the search rate.


    Check the data type of search_conducted to confirm that it's a Boolean Series.
    Calculate the search rate by counting the Series values and expressing them as proportions.
    Calculate the search rate by taking the mean of the Series. (It should match the proportion of True values calculated above.)


In [21]:
import numpy as np

In [22]:
print(ri['search_conducted'].dtype)

bool


In [23]:
ri['search_conducted'].isnull().sum()

0

In [24]:
ri['search_conducted'].value_counts(normalize=True)

False    0.961785
True     0.038215
Name: search_conducted, dtype: float64

In [25]:
ri.search_conducted.mean()

0.0382153092354627

Great! It looks like the search rate is about 3.8%. Next, you'll examine whether the search rate varies by driver gender.

Comparing search rates by gender

In this exercise, you'll compare the rates at which female and male drivers are searched during a traffic stop. Remember that the vehicle search rate across all stops is about 3.8%.

First, you'll filter the DataFrame by gender and calculate the search rate for each group separately. Then, you'll perform the same calculation for both genders at once using a .groupby().

Filter the DataFrame to only include female drivers, and then calculate the search rate by taking the mean of search_conducted.

In [26]:
ri[ri['driver_gender']=='F'].search_conducted.mean()

0.019180617481282074

Filter the DataFrame to only include male drivers, and then repeat the search rate calculation.

In [27]:
ri[ri['driver_gender']=='M'].search_conducted.mean()

0.04542557598546892

Group by driver gender to calculate the search rate for both groups simultaneously. (It should match the previous results.)

In [28]:
ri.groupby(['driver_gender']).search_conducted.mean()

driver_gender
F    0.019181
M    0.045426
Name: search_conducted, dtype: float64

Wow! Male drivers are searched more than twice as often as female drivers. Why might this be?

Adding a second factor to the analysis

Even though the search rate for males is much higher than for females, it's possible that the difference is mostly due to a second factor.

For example, you might hypothesize that the search rate varies by violation type, and the difference in search rate between males and females is because they tend to commit different violations.

You can test this hypothesis by examining the search rate for each combination of gender and violation. If the hypothesis was true, you would find that males and females are searched at about the same rate for each violation. Find out below if that's the case!

Use a .groupby() to calculate the search rate for each combination of gender and violation. Are males and females searched at about the same rate for each violation?

In [29]:
ri.groupby(['driver_gender', 'violation'], as_index=False).search_conducted.mean()

driver_gender            violation  search_conducted
0              F            Equipment          0.039984
1              F     Moving violation          0.039257
2              F                Other          0.041018
3              F  Registration/plates          0.054924
4              F            Seat belt          0.017301
5              F             Speeding          0.008309
6              M            Equipment          0.071496
7              M     Moving violation          0.061524
8              M                Other          0.046191
9              M  Registration/plates          0.108802
10             M            Seat belt          0.035119
11             M             Speeding          0.027885

Reverse the ordering to group by violation before gender. The results may be easier to compare when presented this way.

In [30]:
ri.groupby(['violation', 'driver_gender']).search_conducted.mean()

violation            driver_gender
Equipment            F                0.039984
                     M                0.071496
Moving violation     F                0.039257
                     M                0.061524
Other                F                0.041018
                     M                0.046191
Registration/plates  F                0.054924
                     M                0.108802
Seat belt            F                0.017301
                     M                0.035119
Speeding             F                0.008309
                     M                0.027885
Name: search_conducted, dtype: float64

Great work! For all types of violations, the search rate is higher for males than for females, disproving our hypothesis.

Counting protective frisks

During a vehicle search, the police officer may pat down the driver to check if they have a weapon. This is known as a "protective frisk."

In this exercise, you'll first check to see how many times "Protective Frisk" was the only search type. Then, you'll use a string method to locate all instances in which the driver was frisked.


    Count the search_type values to see how many times "Protective Frisk" was the only search type.
    Create a new column, frisk, that is True if search_type contains the string "Protective Frisk" and False otherwise.
    Check the data type of frisk to confirm that it's a Boolean Series.
    Take the sum of frisk to count the total number of frisks.



In [35]:
ri['search_type'].value_counts(dropna=False)

NaN                                                         83229
Incident to Arrest                                           1290
Probable Cause                                                924
Inventory                                                     219
Reasonable Suspicion                                          214
Protective Frisk                                              164
Incident to Arrest,Inventory                                  123
Incident to Arrest,Probable Cause                             100
Probable Cause,Reasonable Suspicion                            54
Incident to Arrest,Inventory,Probable Cause                    35
Probable Cause,Protective Frisk                                35
Incident to Arrest,Protective Frisk                            33
Inventory,Probable Cause                                       25
Protective Frisk,Reasonable Suspicion                          19
Incident to Arrest,Inventory,Protective Frisk                  18
Incident t

In [37]:
ri['frisk'] = ri['search_type'].str.contains('Protective Frisk', na=False)

In [38]:
ri['frisk'].value_counts()

False    86233
True       303
Name: frisk, dtype: int64

In [39]:
print(ri['frisk'].dtype)

bool


In [41]:
ri['frisk'].sum()

303

Nice job! It looks like there were 303 drivers who were frisked. Next, you'll examine whether gender affects who is frisked.

Comparing frisk rates by gender

In this exercise, you'll compare the rates at which female and male drivers are frisked during a search. Are males frisked more often than females, perhaps because police officers consider them to be higher risk?

Before doing any calculations, it's important to filter the DataFrame to only include the relevant subset of data, namely stops in which a search was conducted.


    Create a DataFrame, searched, that only contains rows in which search_conducted is True.
    Take the mean of the frisk column to find out what percentage of searches included a frisk.
    Calculate the frisk rate for each gender using a .groupby().


In [50]:
searched = ri[ri['search_conducted']==True]

In [51]:
print(searched['frisk'].mean())

0.09162382824312065


In [52]:
searched.groupby('driver_gender').frisk.mean()

driver_gender
F    0.074561
M    0.094353
Name: frisk, dtype: float64

Interesting! The frisk rate is higher for males than for females, though we can't conclude that this difference is caused by the driver's gender.